crates/bevy_openxr/src/interaction.rs

```diff
@@ -0,0 +1,493 @@
+        binding(RIGHT_GRIP_POSE, "/user/hand/right/input/grip/pose"),
+        binding(LEFT_AIM_POSE, "/user/hand/left/input/aim/pose"),
+        binding(RIGHT_AIM_POSE, "/user/hand/right/input/aim/pose"),
+    ]
+}
+
+pub(crate) fn create_actions(
+    instance: &xr::Instance,
+    action_set: &xr::ActionSet,
+    bindings: &[OpenXrBindingDesc],
+) -> HashMap<&'static str, XrAction> {
+    // filter out duplicates, retain the last occurrence
+    let mut filtered_bindings = HashMap::new();
+    for binding in bindings {
+        filtered_bindings.insert(binding.name, binding);
+    }
+
+    let mut actions = HashMap::new();
+    let mut bindings_by_profile = HashMap::<&'static str, Vec<_>>::new();
+    for binding in filtered_bindings.values() {
+        let action = match binding.action_type {
+            OpenXrActionType::BinaryInput => XrAction::BinaryInput(
+                action_set
+                    .create_action(binding.name, binding.name, &[])
+                    .unwrap(),
+            ),
+            OpenXrActionType::FloatInput => XrAction::FloatInput(
+                action_set
+                    .create_action(binding.name, binding.name, &[])
+                    .unwrap(),
+            ),
+            OpenXrActionType::Vector2Input => XrAction::Vector2Input(
+                action_set
+                    .create_action(binding.name, binding.name, &[])
+                    .unwrap(),
+            ),
+            OpenXrActionType::PoseInput => XrAction::PoseInput(
+                action_set
+                    .create_action(binding.name, binding.name, &[])
+                    .unwrap(),
+            ),
+            OpenXrActionType::VibrationOutput => XrAction::VibrationOutput(
+                action_set
+                    .create_action(binding.name, binding.name, &[])
+                    .unwrap(),
+            ),
+        };
+        actions.insert(binding.name, action);
+
+        for id in &binding.paths {
+            bindings_by_profile
+                .entry(id.profile)
+                .or_default()
+                .push((binding.name, id.path));
+        }
+    }
+
+    for (profile, bindings) in bindings_by_profile {
+        let profile = instance.string_to_path(profile).unwrap();
+
+        let mut xr_bindings = vec![];
+        for (name, path) in bindings {
+            let path = instance.string_to_path(path).unwrap();
+
+            let action = actions.get(name).unwrap();
+
+            let binding = match action {
+                XrAction::BinaryInput(action) => xr::Binding::new(action, path),
+                XrAction::FloatInput(action) => xr::Binding::new(action, path),
+                XrAction::Vector2Input(action) => xr::Binding::new(action, path),
+                XrAction::PoseInput(action) => xr::Binding::new(action, path),
+                XrAction::VibrationOutput(action) => xr::Binding::new(action, path),
+            };
+
+            xr_bindings.push(binding);
+        }
```

Comment:
This should be reduced to a `.map().collect()`
```rust
let xr_bindings = bindings.into_iter().map(|(name, path)| /* logic here */).collect();
```

Reply:
When I have big bodies of code I prefer a for loop than map(), but in this case I'm ok with using map().

Reply:
Well it's also a bit of an optimization since calling `push` in a loop usually has worse performance than chaining iterators and then calling `collect` a the end.

Reply:
And it also allows `xr_bindings` to not be `mut` as (imo) the less `mut` variable you have the better 🙃

Reply:
Yeah, I'm all for avoiding mutability. I have to say this method was really hard to write because of the lifetime requirements of xr::Binding, I just stopped worrying about the shape of the code after I solved all compile errors.

Reply:
That's fair :)
Im just trying to catch some of these smaller more rust/code-quality things now before the PR gets too large.